### ISSUE TITLE
Duplicate 3D point for gp4ps

### ISSUE THREAD

**Issue by vlarsson** [vlarsson/PoseLib#12]
The gp4ps implementation does not currently handle having a duplicated 3D point, which could happen in practice. If two 3D points are duplicate, the problem is degenerate, however, if there is only a single duplicate, I believe the problem is minimal with the 4 point correspondences and the solver should work for this as well. In this case it is not possible to filter solutions, so we will have to return multiple solutions.

Doing something like
```
if ((X[0] - X[1]).squaredNorm() < 1e-10) {
        A.row(0).swap(A.row(6));
        A.row(1).swap(A.row(7));
        filter_solutions = false;
 }
```
works if it is the first two points, but we should handle all cases combinations in a smart way. Potentially also detect the degenerate cases and `return 0`.

**Comment by vlarsson** [comment]
Another approach is to implement the solvers from
Camposeco et al., Minimal Solvers for Generalized Pose and Scale Estimation from Two Rays and One Point, ECCV 2016
which should be significantly faster.

**Comment by tsattler** [comment]
I think having an implementation of that solver would be great. 
There could then be a wrapper that calls either of the two solvers depending on the condition. I would like to keep the original gp4ps solver though as to give users the freedom of choice.

**Comment by vlarsson** [comment]
I have implemented the solver from Camposeco et al (see afb4a4b9ce96ed64c6299a24add4ea8ad101a71d). The default interface gp4ps() now checks for duplicates in the 3D points and then either call gp4ps_kukelova or gp4ps_camposeco.

This should fix the problem. I will close the issue, but I guess there is a similar problem for upg3ps (or perhaps other upright solvers). I will need to think about whether this case is also non-degenerate.
